### ISSUE TITLE
unit test failures on Mac OS X 10.10.4/Xcode 6.4

### ISSUE THREAD

**Issue by jeffdiclemente** [CppMicroServices/CppMicroServices#56]
Two unit tests fail (usBundleManifestTest and usLDAPFilterTest ) due to what appears to be how Apple clang handles comparing types.

It appears that Apple clang compares the object addresses and not the string representaiton of the type.

**Comment by saschazelzer** [comment]
This keeps coming back... :-(

Commit 611f732324912b77c7a5c971e5ec92433daf277f removed a workaround for this (also needed for older gcc versions). Look for `US_GCC_RTTI_WORKAROUND_NEEDED`. I guess I was too eager to remove this as I never felt really comfortable with this workaround.

Could you provide the versions of clang, ld and libc++ you are using?

Could you also search with `nm -om` (I think) through all object files for `weak private external typeinfo` entries that have a corresponding typeinfo entry in another object file, but with `weak external typeinfo`?

I found [1] a valuable blog post for this kind of issue.

[1] http://www.russellmcc.com/posts/2013-08-03-rtti.html

**Comment by jeffdiclemente** [comment]
Thanks for the information, that's helpful.


as for the versions...

% xcrun clang --version
Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.4.0
Thread model: posix

% xcrun ld -v
@(#)PROGRAM:ld  PROJECT:ld64-242.2
configured to support archs: armv6 armv7 armv7s arm64 i386 x86_64 x86_64h armv6m armv7m armv7em
LTO support using: LLVM version 3.6.0svn

not sure what the exact version of libc++ is, but it'll be whichever ships with Xcode 4.6.0.

I'll take a look through all the object files and let you know what I find.

I've already fixed part of the problem by making the Any class have default visibility. std::map<std::string, Any> is still failing the type comparison though.

**Comment by jeffdiclemente** [comment]
After working through this issue some more, here is what I know.

There is still a puzzle as to why the typeinfo symbols are handled the way they are (read on for details about what I mean).

I'm using Xcode 4.6.0 (clang 6.1.0).

First, I modified the ```Any``` class to have default visibility (i.e. added ```US_Core_EXPORT``` to the declaration) and moved the definitions of non-template functions into a separate source file. I also marked all template functions with ```US_Core_EXPORT```. 
This resolved the ```us::Any``` typeinfo symbol visibility (i.e. weak external typeinfo symbols) across shared library boundaries. 

```std::vector<us::Any>``` and ```std::map<std::string, us::Any>``` still have typeinfo symbol visibility problems; within the same binary (in this case usCoreTestDriver), one compilation unit has the symbol as **weak private external typeinfo** (usBundleManifestTest.o) and others have the symbol as **weak external typeinfo** (LDAPFilterTest.o and usAnyTest.o). I was able to isolate the source of the weak private external symbol in usBundleManifestTest.o to the ```template<class S> std::vector<ServiceReference<S>> GetServiceReferences(const std::string& filter = std::string())``` function in ```usBundleContext.h```. Specifically, the ```BaseVectorT::const_iterator i``` statement in the for loop is causing the weak private external typeinfo symbol. 
I paired down the usCoreTestDriver binary to only build ```usCoreTestDriver.cpp```, ```us_init.cpp``` and ```usBundleManifestTest.cpp``` and simplified this source file to only include:

```
#include <usAny.h>

int usBundleManifestTest(int /*argc*/, char* /*argv*/[])
{
    std::vector<std::basic_string<char>>::const_iterator ci;
    const std::type_info& t1 = typeid(std::vector<us::Any>);
    (void)t1;
   return 0;
}
```

Building this code produces a usBundleManifestTest.o with a **weak private external typeinfo symbol** for ```std::vector<us::Any>```. As a side note, changing ```std::vector<std::basic_string<char>>::const_iterator ci;``` to be a vector of type _char\*, int, us::Bundle, std::shared_ptr<us::Bundle>, void* or std::string_ produces the same **weak private external typeinfo symbol** for ```std::vector<us::Any>```. I'd guess that any type will produce this behavior, not just the ones I've listed.

The interesting part; moving ```std::vector<std::basic_string<char>>::const_iterator ci;``` to after the ```(void)t1;```statement produces a usBundleManifestTest.o with a **weak external typeinfo symbol** for ```std::vector<us::Any>```.
I used clang's ```-ast-dump``` feature to see the AST of usBundleManifest.cpp. I can see that there are differences however I'm not sure what they mean yet.
[The two ASTs are here](https://github.com/CppMicroServices/CppMicroServices/files/144638/usBundleManifestTest-AST.zip)

What i've been unable to figure out is; why is the typeinfo symbol for ```std::vector<us::Any>``` being modified by the inclusion of a type which is not ```std::vector<us::Any>```?

Also, I did all of the above using Xcode 6.2.0 (clang 6.0) and the results are exactly the same as above.

Some next steps:
- I'm going to try reverting back to the ```US_GCC_RTTI_WORKAROUND_NEEDED``` workaround to see whether that does anything. 
- Is this a clang issue (I can try older versions of clang)? At the moment I have no evidence to suggest that it is.
- Another possible workaround is to modify ```template <typename ValueType> ValueType* any_cast(Any* operand)``` to compare the type's name instead of the type_info object pointers. This will be slower and is really not actually fixing the fault.

**Comment by jeffdiclemente** [comment]
I believe there is a bug in clang and I've filed [a bug report with clang](https://llvm.org/bugs/show_bug.cgi?id=26749).

**Comment by jeffdiclemente** [comment]
closing as fixed in PR #82
